### PR TITLE
Key cache

### PR DESCRIPTION
**Pull request by cachedout**
### What does this PR do?

Implements an optional master-side metadata cache for keys
### What issues does this PR fix or reference?

None
### Previous Behavior

The master would have to open O(n) files for the number of accepted keys.
### New Behavior

Much closer to O(1). This should be a substantial increase for masters with tens of thousand or hundreds of thousands of accepted keys.

### CODE REVIEW

**Comment by s0undt3ch** [comment]
Maybe add key add/delete/reject/etc event listening to the maintenance process to update the cache before the next iteration? This would keep it always in sync?

**Comment by cachedout** [comment]
@s0undt3ch Absolutely! This is actually why I gave the config option a string type. I'm in the process of writing an engine to do exactly what you suggest.

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
Go Go Jenkins!

File: doc/topics/tutorials/intro_scale.rst [hunk 1]
```diff
@@ -271,3 +271,17 @@ If the job cache is necessary there are (currently) 2 options:
   into a returner (not sent through the Master)
 - master_job_cache (New in `2014.7.0`): this will make the Master store the job
   data using a returner (instead of the local job cache on disk).
+
+If a master has many accepted keys, it may take a long time to publish a job
+because the master much first determine the matching minions and deliver
```
Comment (rallytime) [inline doc/topics/tutorials/intro_scale.rst]: `must`

**Comment by cachedout** [comment]
Reviewed by @terminalmage. Merging after adding docs, per his suggestion.

**Comment by terminalmage** [comment]
:+1:

**Pull request merged**